pkg/vault/vaultfactory.go

```diff
@@ -0,0 +1,78 @@
+package vault
+
+import (
+	"github.com/hashicorp/vault/api"
+	"github.com/jenkins-x/jx/pkg/jx/cmd/common"
+	"github.com/jenkins-x/jx/pkg/kube/serviceaccount"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type VaultClientFactory struct {
+	Options          common.OptionsInterface
+	Selector         VaultSelector
+	kubeClient       kubernetes.Interface
+	defaultNamespace string
+}
+
+func NewVaultClientFactory(options common.OptionsInterface) (VaultClientFactory, error) {
+	factory := VaultClientFactory{
+		Options: options,
+	}
+	var err error
+	factory.kubeClient, factory.defaultNamespace, err = options.KubeClient()
+	if err != nil {
+		return factory, err
+	}
+	factory.Selector, err = NewVaultSelector(options)
+	if err != nil {
+		return factory, err
+	}
+	return factory, nil
+}
+
+// NewVaultClient creates a new api.Client
+// if namespace is nil, then the default namespace of the factory will be used
+// if the name is nil, and only one vault is found, then that vault will be used. Otherwise the user will be prompted to
+// select a vault for the client.
+func (v VaultClientFactory) NewVaultClient(name string, namespace string) (*api.Client, error) {
+	config, jwt, role, err := v.GetConfigData(name, namespace)
+	if err != nil {
+		return nil, err
+	}
+	vaultClient, err := api.NewClient(config)
+	token, err := getTokenFromVault(role, jwt, vaultClient)
+	vaultClient.SetToken(token)
+	return vaultClient, nil
+}
+
+// GetConfigData generates the information necessary to configure an api.Client object
+// Returns the api.Config object, the JWT needed to create the auth user in vault, and an error if present
+func (v *VaultClientFactory) GetConfigData(name string, namespace string) (config *api.Config, jwt string, saName string, err error) {
```

Comment:
I think we can make this function private to the package by renaming to `getConfigData`.

Reply:
It's used in the _test package, so it can't be :-(